README.md
# Cypher Kernel

This is a small Jupyter kernel wrapping the Cypher language and Neo4j [https://neo4j.com/developer/cypher/](https://neo4j.com/developer/cypher/).

![](docs/what_is_this.gif)


## Why? Do I need a Cypher language kernel?

I started working on this kernel when I was teaching Cypher and Neo4j as part of a database course at Copenhagen Business Academy.

Usually, my lecture notes are in Jupyter notebooks and I use the Jupter extension [RISE](https://github.com/damianavila/RISE) to make slideshows with executable code.

In the classes on graph databases, I wanted to focus on the query language Cypher. That is, I do not want to have Cypher embedded in Python such as with [icypher](https://github.com/lebedov/icypher) -a `%cypher` magic sending queries with the help of `py2neo`- or in [plain Python notebooks](https://nicolewhite.github.io/neo4j-jupyter/hello-world.html)


## How does it look like?

In case you do not have a Python environment with Jupyter installed on your system, you can see in the following how the rendered notebooks -not the presentations- could look like, see:

  * https://nbviewer.jupyter.org/github/HelgeCPH/cypher_kernel/blob/master/example/paradise_papers.ipynb
  * https://nbviewer.jupyter.org/github/HelgeCPH/cypher_kernel/blob/master/example/Mark_Needhams_Examples.ipynb
  * https://nbviewer.jupyter.org/github/HelgeCPH/cypher_kernel/blob/master/example/movie_graph.ipynb or
  * http://htmlpreview.github.io/?https://github.com/HelgeCPH/cypher_kernel/blob/master/example/movie_graph.html 



## Installation

To install the `cypher_kernel` from PyPi:

```bash
pip install cypher_kernel
python -m cypher_kernel.install
```

To work on this code directly, you may want to:

```bash
git clone git@github.com:HelgeCPH/cypher_kernel.git
cd cypher_kernel
pip install .
python -m cypher_kernel.install
```

### Dependencies

## Configuration

To configure a Neo4j user, password, and the address of the Neo4j REST API, you can specify the values in the configuration file `cypher_config.yml`. Normally, this file is located under `~/.jupyter/`. In case the file is not existent the following default configuration is used:

```yaml
user: 'neo4j'
pwd: 'pwd'
host: 'localhost:7474'
connect_result_nodes: False
cmd_timeout: null
```


## Using the Cypher Kernel

**Notebook**: The *New* menu in the notebook should show an option for an `Cypher` notebook.

**Console frontends**: To use it with the console frontends, add `--kernel cypher` to their command line arguments.



## Neo4j for Presentations


To get quickly started -under the assumption you have Docker installed- start up a Neo4j DBMS instance with: 

```bash
docker run \
    --rm \
    --publish=7474:7474 \
    --publish=7687:7687 \
    --env NEO4J_AUTH=neo4j/pwd \
    neo4j
```

  * After the DB engine started, create a new Cypher notebook.

See more on configuring Neo4j Docker containers [here](https://neo4j.com/docs/operations-manual/current/installation/docker/)


### What? I have Docker but no `pip` and other Python stuff?!

Likely the easiest way to get started, have:

  * A Linux/OS X (Windows should work but I cannot test it at the moment...)
  * An installation of Anaconda (with Python 3.6 or higher), see https://www.anaconda.com/download/. Download and install it according to their documentation, see https://docs.anaconda.com/anaconda/install/
  * A Docker installation, see https://www.docker.com/community-edition#/download. Alternatively, a native installation of Neo4j, see https://neo4j.com/download/
  * The `cypher_kernel`:
  ```bash
  pip install cypher_kernel
  python -m cypher_kernel.install
  ```
  * The Jupyter Notebook server up and running:
  ```bash
  jupyter notebook
  ```

That should be it...




CHANGELOG.md
## Changes between `cypher_kernel` 0.3.1 and 0.3.2

  * Fixed a dependency issue in `setup.py`, which prevented installation of 
    dependencies before their use.

## Changes between `cypher_kernel` 0.3.0 and 0.3.1

  * Gave the project a proper long description for PyPI in rst format.

## Changes between `cypher_kernel` 0.2.3 and 0.3.0

  * Removed dependency to and wrapping of `cypher-shell`
  * Switched to Neo4j Python driver instead
  * Textual output looks a bit different than earlier, 
    `example/Cypher Kernel Example.ipynb` and `example/movie_graph.ipynb` have 
    been updated accordingly
  

## Changes between `cypher_kernel` 0.2.2 and 0.2.3

  * Fixed handling of queries, which do not return anything, such as the 
    following. Earlier they crashed the kernel, now they should work.
  ```
  MERGE (nAlice:User {id:'Alice'})
  MERGE (nBridget:User {id:'Bridget'})
  MERGE (nCharles:User {id:'Charles'})
  MERGE (nDoug:User {id:'Doug'})
  MERGE (nMark:User {id:'Mark'})
  MERGE (nMichael:User {id:'Michael'})
  MERGE (nAlice)-[:FOLLOWS]->(nDoug)
  MERGE (nAlice)-[:FOLLOWS]->(nBridget)
  MERGE (nAlice)-[:FOLLOWS]->(nCharles)
  MERGE (nMark)-[:FOLLOWS]->(nDoug)
  MERGE (nMark)-[:FOLLOWS]->(nMichael)
  MERGE (nBridget)-[:FOLLOWS]->(nDoug)
  MERGE (nCharles)-[:FOLLOWS]->(nDoug)
  MERGE (nMichael)-[:FOLLOWS]->(nDoug)
  ```


## Changes between `cypher_kernel` 0.2.1 and 0.2.2

  * Added handling for `EXPLAIN` and parsing for `PROFILE` queries.


## Changes between `cypher_kernel` 0.2.0 and 0.2.1

  * Fixed parsing of queries, such as `CALL apoc.meta.graph();`, which may 
    return lists containing only single elements, as in the following:

  ```
  +-----------------------+
  | nodes | relationships |
  +-----------------------+
  | [(:Person {name: "Person", count: 737000, _id_: -12})] | [[:FOLLOWS {count: 36733633, _id_: -12, type: "FOLLOWS", in: 36733633, out: 36733633}[-12>-12]]] |
  +-----------------------+
  ```

  * Fixed parsing of relations, which was broken due to a mistake in parsing 
    lists.
  * Added keywords for `EXPLAIN` and `PROFILE`, however their results currently 
    crash the parser.




## Changes between `cypher_kernel` 0.1.1 and 0.2.0

### Added parsing of lists

Cypher queries, may return lists. Examples are:

```cypher
CALL apoc.meta.graph();
```

```cypher
MATCH (n) 
RETURN labels(n) AS labels, COUNT(*) AS count
ORDER BY count DESC
```

Such lists should now be parsed correctly.


### Added support for `%%python` magic

Now, there are standard code cells, which contains Cypher code and code cells 
for `%%bash` and `%%python` code.

Noteable about `%%python` code cells is, that a Pandas `DataFrame` is 
accessible under the name `df`. It contains the results of the last Cypher 
query. That is, the `DataFrame` may be empty of the query did not return 
anything. Otherwise, it contains the data as returned by the query. Note, all 
contents are strings and have to be manually casted correspondingly.

For example, if a query returns:

```
+------------------------------------------------------+
| juris            | num   | officer_country           |
+------------------------------------------------------+
| "Bermuda"        | 67025 | "United States"           |
| "Aruba"          | 36191 | "Netherlands Antilles"    |
| "Bermuda"        | 12846 | "United Kingdom"          |
| "Bermuda"        | 12180 | "Hong Kong"               |
| "Cayman Islands" | 9563  | "United States"           |
| "Bermuda"        | 7818  | "Switzerland"             |
| "Aruba"          | 7168  | "Curaçao"                 |
| "Aruba"          | 5344  | "Virgin Islands, British" |
| "Bermuda"        | 4089  | "China"                   |
| "Bermuda"        | 3794  | "Canada"                  |
| "Aruba"          | 3627  | "Saint Kitts and Nevis"   |
| "Bermuda"        | 3123  | "Singapore"               |
| "Cayman Islands" | 2585  | "China"                   |
| "Bermuda"        | 2556  | "Australia"               |
| "Bermuda"        | 2219  | "British Virgin Islands"  |
| "Isle of Man"    | 2032  | "United Kingdom"          |
| "Bermuda"        | 1912  | "Cayman Islands"          |
| "Cayman Islands" | 1739  | "United Kingdom"          |
| "Bermuda"        | 1629  | "France"                  |
| "Cayman Islands" | 1596  | "Hong Kong"               |
+------------------------------------------------------+
```

The corresponding `DataFrame` looks as:

```
               juris    num            officer_country
0          "Bermuda"  67025            "United States"
1            "Aruba"  36191     "Netherlands Antilles"
2          "Bermuda"  12846           "United Kingdom"
3          "Bermuda"  12180                "Hong Kong"
4   "Cayman Islands"   9563            "United States"
5          "Bermuda"   7818              "Switzerland"
6            "Aruba"   7168                  "Curaçao"
7            "Aruba"   5344  "Virgin Islands, British"
8          "Bermuda"   4089                    "China"
9          "Bermuda"   3794                   "Canada"
10           "Aruba"   3627    "Saint Kitts and Nevis"
11         "Bermuda"   3123                "Singapore"
12  "Cayman Islands"   2585                    "China"
13         "Bermuda"   2556                "Australia"
14         "Bermuda"   2219   "British Virgin Islands"
15     "Isle of Man"   2032           "United Kingdom"
16         "Bermuda"   1912           "Cayman Islands"
17  "Cayman Islands"   1739           "United Kingdom"
18         "Bermuda"   1629                   "France"
19  "Cayman Islands"   1596                "Hong Kong"
```


### Syntax highlighting

Cypher code is now highlighted correctly. Thanks to Thomas' answer to my 
question on the Jupyter Help repository, see 
https://github.com/jupyter/help/issues/301

cypher_kernel/kernel.py
import os
import sys
import json
import yaml
import uuid
import random
import platform
from jinja2 import Template
from neo4j.data import Node, Relationship
from neo4j.exceptions import Neo4jError
from ipykernel.kernelbase import Kernel
from .cypher_keywords import KEYWORDS
from .neo4j_connection import Neo4jConnection
from pexpect.replwrap import bash, python


class CypherKernel(Kernel):
    implementation = "Cypher"
    implementation_version = sys.modules[__package__].__version__
    language = "cypher"
    language_version = "0.1"
    language_info = {
        # Switched that to cypher, see
        # https://github.com/jupyter/help/issues/301
        "name": "cypher",
        "mimetype": "text/cypher",
        "file_extension": ".cql",
    }
    banner = "Cypher kernel - Neo4j in Jupyter Notebooks"
    keywords = KEYWORDS

    global_node_colors = {}

    my_python = python(command="python")
    my_python.run_command(
        """import ast
import pandas as pd
import networkx as nx
G = nx.Graph()
df = pd.DataFrame()"""
    )

    if platform.system() == "Windows":
        # TODO: what shall I do on Windows here???
        # my_shell = ...
        pass
    else:
        my_shell = bash(command="bash")

    @property
    def cfg(self):
        cfg = CypherKernel._parse_config()
        return cfg

    @property
    def user(self):
        return self.cfg["user"]

    @property
    def pwd(self):
        return self.cfg["pwd"]

    @property
    def host(self):
        return self.cfg["host"]

    @property
    def connect_result_nodes(self):
        return self.cfg["connect_result_nodes"]

    @property
    def cmd_timeout(self):
        return self.cfg["cmd_timeout"]

    def __init__(self, **kwargs):
        Kernel.__init__(self, **kwargs)
        # establish connection to DB
        self.conn = Neo4jConnection(uri=self.host, user=self.user, pwd=self.pwd)

    @staticmethod
    def _parse_config():
        """
        Parses spawn YAML command options from the default Jupyter configuration directory.
        http://jupyter.readthedocs.io/en/latest/projects/jupyter-directories.html#configuration-files
        """
        config_dir = os.environ.get("JUPYTER_CONFIG_DIR")
        if not config_dir:
            config_dir = ".jupyter"
        config_path = os.path.join(os.path.expanduser("~"), config_dir, "cypher_config.yml")
        default_config = {
            "user": "neo4j",
            "pwd": "pwd",
            "host": "neo4j://localhost:7687",
            "connect_result_nodes": False,
            "cmd_timeout": None,
        }
        try:
            config = yaml.load(open(config_path), Loader=yaml.FullLoader)
            if len([k for k in default_config.keys() if k in config.keys()]) == 5:
                return config
        except FileNotFoundError:
            # Using default configuration
            return default_config
        return default_config

    def _response_to_js_graph(self, nodes, relations, element_id):
        template_str = """require(["https://cdnjs.cloudflare.com/ajax/libs/vis/4.21.0/vis.min.js"], function(vis) {
  var nodes = new vis.DataSet([
    {% for n in nodes %}{ id: {{ n.id }}, label: "{{ label_picker(n.labels) }}", title: "{{ properties_serializer(n._properties) }}", color: 'rgba({{ node_colors[label_picker(n.labels)] }})'},
    {% endfor %}
  ]);

  // create an array with edges
  var edges = new vis.DataSet([
    {% for r in rels %}{ from: {{ r.nodes[0].id }},to: {{ r.nodes[1].id }}, arrows:'to', title: "{{ r.type }}" },
    {% endfor %}
  ]);

  // create a network
  var container = document.getElementById('{{element_id}}');
  var data = {
    nodes: nodes,
    edges: edges
  };

  var options = {
    edges: {
      arrows: {
        to: {
          scaleFactor: 0.5
        }
      }
    },
    width: '100%',
    height: '500px',
    interaction: {hover: true}
  };

  var network = new vis.Network(container, data, options);});
"""

        def label_picker(frozenset):
            """This function is only necessary, since frozensets cannot be easily cast to lists in Jinja2"""
            return list(frozenset)[0]

        def escape_strs(properties_dict):
            properties_str = "{"
            for k, v in properties_dict.items():
                if type(v) == str and "'" in v:
                    v.replace("'", "'")
                    # v.replace('"', '\"')
                properties_str += k + ":" + str(v) + ",<br>"
            properties_str = properties_str[:-5] + "}"
            return properties_str

        template = Template(template_str)
        graph_js = template.render(
            nodes=nodes,
            rels=relations,
            element_id=element_id,
            node_colors=self.global_node_colors,
            label_picker=label_picker,
            properties_serializer=escape_strs,
        )

        with open("/tmp/out.js", "w") as fp:
            fp.write(graph_js)
        return graph_js

    def _color_nodes(self, nodes):
        for n in nodes:
            labels_lst = list(n.labels)
            if not labels_lst[0] in self.global_node_colors.keys():
                rgb = [str(random.randint(0, 255)) for _ in range(3)]
                rgba = rgb + ["0.5"]
                self.global_node_colors[labels_lst[0]] = ",".join(rgba)

    def _clean_input(self, code):
        lines = code.splitlines()
        clean_input = "\n".join([l for l in lines if l])
        return clean_input

    def _is_magic(self, code):
        magic_lines = code.splitlines()
        magic_line = magic_lines[0].strip()
        if magic_line.startswith("%%"):
            return magic_line.replace("%%", ""), "\n".join(magic_lines[1:])
        else:
            return None, None

    def _send_to_bash(self, code):
        res = self.my_shell.run_command(code, timeout=self.cmd_timeout)
        return res

    def _send_to_python(self, code):
        res = self.my_python.run_command(code, timeout=self.cmd_timeout)
        return res

    def _push_to_python_env(self, header, content):
        code = '''header = ast.literal_eval("""{}""")
content = ast.literal_eval("""{}""")
df = pd.DataFrame(content, columns=header)
df'''.format(
            str(header).replace("'", "\\'"), str(content).replace("'", "\\'")
        )
        res = self._send_to_python(code)
        return res

    def _construct_and_send_text_response(self, response, is_silent, status="ok"):
        if not is_silent:
            result = {"data": {"text/plain": response}, "execution_count": self.execution_count}
            self.send_response(self.iopub_socket, "execute_result", result)

        exec_result = {
            "status": status,
            "execution_count": self.execution_count,
            "payload": [],
            "user_expressions": {},
        }
        return exec_result

    def _construct_and_send_html_response(self, nodes, relations, is_silent, status="ok"):
        element_id = uuid.uuid4()
        graph_js = self._response_to_js_graph(nodes, relations, element_id)

        graph_HTML_tmpl = """<link href="https://cdnjs.cloudflare.com/ajax/libs/vis/4.21.0/vis.min.css" rel="stylesheet" type="text/css">
        <div id="{{ element_id }}"></div>
        """
        html_template = Template(graph_HTML_tmpl)
        graph_HTML = html_template.render(element_id=element_id)

        html_msg = {"data": {"text/html": graph_HTML}, "execution_count": self.execution_count}
        js_msg = {"data": {"application/javascript": graph_js}}
        self.send_response(self.iopub_socket, "display_data", js_msg)
        self.send_response(self.iopub_socket, "display_data", html_msg)

        exec_result = {"status": status, "execution_count": self.execution_count, "payload": [], "user_expressions": {}}
        return exec_result

    def _process_response(self, response):
        """Split nodes and relations from response records for later visualization"""
        nodes, relations = set([]), set([])
        text_response = ""
        for record in response:
            for el in record:
                if isinstance(el, Node):
                    nodes.add(el)
                elif isinstance(el, Relationship):
                    # Then it is a relation wrapped into an abstract base class
                    relations.add(el)
                else:
                    text_response += str(el) + "\n"
        return nodes, relations, text_response

    def do_execute(self, code, silent, store_history=True, user_expressions=None, allow_stdin=False):

        clean_input = self._clean_input(code)
        magic, magic_code = self._is_magic(code)
        if magic == "bash" and magic_code:
            response = self._send_to_bash(magic_code)
            exec_result = self._construct_and_send_text_response(response, silent)
            return exec_result
        elif magic == "python":
            # TODO! implement me, convert latest cypher query to networkx graph
            response = self._send_to_python(magic_code)
            exec_result = self._construct_and_send_text_response(response, silent)
            return exec_result
        elif magic == None:
            # then it is a query to Cypher
            try:
                # Run the actual cypher query
                response = self.conn.query(code)
            except Neo4jError as e:
                # Send an error message to the text output
                exec_result = self._construct_and_send_text_response(e.message, silent, status="error")
                return exec_result

            nodes, relations, text_response = self._process_response(response)

            # TODO: push results to python env via to_df() method on results
            # df_header, df_content = parse_output_to_python(text_response)
            # if df_header and df_content:
            #     _ = self._push_to_python_env(df_header, df_content)

            if not silent and nodes:
                # Only return the visual output when there are actually nodes and relations,
                # as long as auto connection is not implemented also put it there when only nodes exist
                self._color_nodes(nodes)
                exec_result = self._construct_and_send_html_response(nodes, relations, silent, status="ok")
                return exec_result
            if not silent:
                # No matter what, this is the text response
                exec_result = self._construct_and_send_text_response(text_response, silent)
                return exec_result
        else:
            response = f"Unknown magic type: {magic}"
            exec_result = self._construct_and_send_text_response(response, silent, status="error")
            return exec_result
        # TODO: What should the exec_result be in case of silence?
        return None

    def do_complete(self, code, cursor_pos):
        space_idxs = [i for i, l in enumerate(code) if l == " "]
        low_idxs = [s for s in space_idxs if s < cursor_pos]
        if low_idxs:
            low_cp = max([s for s in space_idxs if s < cursor_pos]) + 1
            key_start = code[low_cp:cursor_pos]
        else:
            low_cp = 0
            key_start = code[:cursor_pos]

        matches = [k for k in self.keywords if k.startswith(key_start)]
        content = {"matches": matches, "cursor_start": low_cp, "cursor_end": cursor_pos, "metadata": {}, "status": "ok"}
        return content
